### ISSUE TITLE
CONNECTION_LOST not an Integer...

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3455]
|<img alt="mforbes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| mforbes reported|
|-|-|
|Trac ID|trac#3455|
|Type|defect|
|Created|2008-10-01 21:13:36Z|
|Branch|https://github.com/twisted/twisted/tree/writesomedata-docs-3455|

Several docstrings, notably those for the writeSomeData() method of t.i.a.FileDescriptor, claim that the return value is an integer reporting the number of bytes sent, or a negative integer reporting that the connection was lost.

The implementations, however, return t.i.main.CONNECTION_LOST which is an exception, not an integer.

Apart from the incorrect docstrings, returning an exception makes it quite difficult to test for errors.  For example:

```
bytes_sent = self.connection.writeSomeData(...)
if 0 < bytes_sent:
    Still get here, even if connection is lost.
```

One has to check if the return value is explicitly CONNECTION_LOST, or if it is an instance of int etc.

Suggestion: Have writeSomeData throw the exception, or revert back to using an integer...

A quick grep indicates that this would require changes in the following files.

./Twisted/twisted/internet/_javaserialport.py
./Twisted/twisted/internet/_posixserialport.py
./Twisted/twisted/internet/abstract.py
./Twisted/twisted/internet/fdesc.py
./Twisted/twisted/internet/process.py
./Twisted/twisted/internet/tcp.py
./Twisted/twisted/protocols/htb.py
./Twisted/twisted/protocols/pcp.py
./Twisted/twisted/test/test_internet.py
./Twisted/twisted/test/test_pcp.py

<details><summary>Searchable metadata</summary>

```
trac-id__3455 3455
type__defect defect
reporter__mforbes mforbes
priority__normal normal
milestone__ 
branch__branches_writesomedata_docs_3455 branches/writesomedata-docs-3455
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1222895616000000 1222895616000000
changetime__1227117677000000 1227117677000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

Sorry, this is a documentation issue, but the behavior won't change, for several reasons (compatibility, performance, and others). Note that the fix is quite simple:

```
bytes_sent = self.connection.writeSomeData(...)
if bytes_sent < 0 or isinstance(bytes_sent, Exception):
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [25420]) Branching to 'writesomedata-docs-3455'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

Turns out, no implementation of `writeSomeData` returns a negative integer to indicate connection loss.

I could include this hunk in the branch if I wanted:

```
#!diff
Index: twisted/internet/abstract.py
===================================================================
--- twisted/internet/abstract.py        (revision 25421)
+++ twisted/internet/abstract.py        (working copy)
@@ -111,7 +111,7 @@
             l = self.writeSomeData(buffer(self.dataBuffer, self.offset))
         else:
             l = self.writeSomeData(self.dataBuffer)
-        if l < 0 or isinstance(l, Exception):
+        if isinstance(l, Exception):
             return l
         if l == 0 and self.dataBuffer:
             result = 0
```

without breaking any of Twisted's unit tests, but I'm not going to because perhaps some third-party writeSomeData implementation decided to take the negative return value documentation to heart and do that instead of returning an exception.

So I've just corrected the documentation to say that writeSomeData may returned an integer or an exception and dropped all the text about negative integers.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_6"></a>|@mwhudson set owner to @exarkun|
|-|-|

Looks fine to me.  Please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

(In [25452]) Merge writesomedata-docs-3455

Author: exarkun
Reviewer: mwhudson
Fixes: [#3455](https://github.com/twisted/twisted/issues/3455)

Change the API documentation for writeSomeData so that it does not claim negative
values are returned to indicate connection lose.  Instead document that an exception
instance will be returned to indicate that.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
